src/stages/models/manager.ts

```diff
@@ -197,6 +197,22 @@ export class StageManager {
       throw new StageNotFoundError(stagesErrorMessages.stageNotFound);
     }
 
+    const previousStageOrder = stage.order - 1;
+
+    // can't move to PENDING if previous stage is not COMPLETED
+    if (status === StageOperationStatus.PENDING && previousStageOrder > 0) {
+      const previousStage = await prisma.stage.findFirst({
+        where: {
+          jobId: stage.jobId,
+          order: previousStageOrder,
+        },
+      });
+
+      if (previousStage?.status !== StageOperationStatus.COMPLETED) {
```

Comment:
You are not handling the case when a stage is not returned (and the previous stage is undefined). If it should not happen, consider using `!` rather than `?`
```suggestion
      if (previousStage!.status !== StageOperationStatus.COMPLETED) {
```